File: typescript/aeon-3-sayonara-sun/src/Farewell.ts
export class Farewell {
  readonly version: string;
  readonly daytime: string;
  readonly farewell: string;
  readonly sleepTight: string;
  readonly sunsetHour: number;
  readonly cooldownHours: number;
  readonly sleepHour: number;
  readonly wakeHour: number;

  constructor(data: {
    version: string;
    daytime: string;
    farewell: string;
    sleepTight: string;
    sunsetHour: number;
    cooldownHours: number;
    sleepHour: number;
    wakeHour: number;
  }) {
    this.version = data.version;
    this.daytime = data.daytime;
    this.farewell = data.farewell;
    this.sleepTight = data.sleepTight;
    this.sunsetHour = data.sunsetHour;
    this.cooldownHours = data.cooldownHours;
    this.sleepHour = data.sleepHour;
    this.wakeHour = data.wakeHour;
  }

  getSunsetWindow(): string {
    const endHour = (this.sunsetHour + this.cooldownHours) % 24;
    return `${String(this.sunsetHour).padStart(2, '0')}:00-${String(endHour).padStart(2, '0')}:00`;
  }

  getSleepWindow(): string {
    return `${String(this.sleepHour).padStart(2, '0')}:00-${String(this.wakeHour).padStart(2, '0')}:00`;
  }

  private hourInWindow(currentHour: number, startHour: number, endHour: number): boolean {
    if (startHour === endHour) return true;
    if (startHour < endHour) return currentHour >= startHour && currentHour < endHour;
    return currentHour >= startHour || currentHour < endHour;
  }

  getMessage(currentHour: number): string {
    const sunsetEnd = (this.sunsetHour + this.cooldownHours) % 24;
    if (this.hourInWindow(currentHour, this.sleepHour, this.wakeHour)) {
      return this.sleepTight;
    }
    if (this.hourInWindow(currentHour, this.sunsetHour, sunsetEnd)) {
      return this.farewell;
    }
    return this.daytime;
  }
}

File: python/python-sayonara-sun/src/farewell.py
from __future__ import annotations

from dataclasses import dataclass


@dataclass(slots=True)
class Farewell:
    version: str
    daytime: str
    farewell: str
    sleep_tight: str
    sunset_hour: int
    cooldown_hours: int
    sleep_hour: int
    wake_hour: int

    def get_sunset_window(self) -> str:
        end_hour = (self.sunset_hour + self.cooldown_hours) % 24
        return f"{self.sunset_hour:02d}:00-{end_hour:02d}:00"

    def get_sleep_window(self) -> str:
        return f"{self.sleep_hour:02d}:00-{self.wake_hour:02d}:00"

    def _hour_in_window(self, current_hour: int, start_hour: int, end_hour: int) -> bool:
        if start_hour == end_hour:
            return True
        if start_hour < end_hour:
            return start_hour <= current_hour < end_hour
        return start_hour <= current_hour or current_hour < end_hour

    def get_message(self, current_hour: int) -> str:
        sunset_end = (self.sunset_hour + self.cooldown_hours) % 24
        if self._hour_in_window(current_hour, self.sleep_hour, self.wake_hour):
            return self.sleep_tight
        if self._hour_in_window(current_hour, self.sunset_hour, sunset_end):
            return self.farewell
        return self.daytime
